contracts/StorageFees.cdc

```diff
@@ -0,0 +1,244 @@
+
+    // Defines the path where each account should have a `StorageReservationReceiver` capability
+    pub let storageReservationReceiverPath: PublicPath
+
+    // Defines the path where each account should have a `StorageReservation` capability
+    pub let storageReservationPath: StoragePath
+
+    // Defines how much storage capacity an account has per reserved Flow token.
+    // definition is written per unit of flow instead of the inverse, so there is no loss of precision calculating storage from flow, but there is loss of precision when calculating flow per storage.
+    pub var storageBytesPerReservedFlow: UFix64
+
+    // Defines the minimum amount of Flow tokens that an account needs to have reserved for storage capacity.
+    // If an account has less then this amount reserved by the end of any transaction it participated in, the transaction will fail.
+    pub var minimumStorageReservation: UFix64
+
+    // Enables or disables the refunding storage capacity.
+    pub var refundingEnabled: Bool
+
+    // An administrator resource that can change the parameters of the StorageFees smart contract.
+    pub resource Administrator {
+        pub fun setRefundingEnabled(_ enabled: Bool) {
+            if StorageFees.refundingEnabled == enabled {
+              return
+            }
+            StorageFees.refundingEnabled = enabled
+            emit RefundingEnabledChanged(enabled)
+        }
+
+        // Changes the amount storage capacity an account has per accounts' reserved storage.
+        pub fun setStorageBytesPerReservedFlow(_ storageBytesPerReservedFlow: UFix64) {
+            if StorageFees.storageBytesPerReservedFlow == storageBytesPerReservedFlow {
+              return
+            }
+            StorageFees.storageBytesPerReservedFlow = storageBytesPerReservedFlow
+            emit StorageBytesPerReservedFlowChanged(storageBytesPerReservedFlow)
+        }
+
+        // Changes the minimum amount of Flow tokens an account has to have reserved.
+        pub fun setMinimumStorageReservation(_ minimumStorageReservation: UFix64) {
+            if StorageFees.minimumStorageReservation == minimumStorageReservation {
+              return
+            }
+            StorageFees.minimumStorageReservation = minimumStorageReservation
+            emit MinimumStorageReservationChanged(minimumStorageReservation)
+        }
+
+        access(contract) init(){}
+    }
+
+    // An interface for public access to accounts' storage reservation.
+    // If `StorageReservationReceiver` capability is not on the accounts' `storageReservationReceiverPath` path
+    // the account won't be able to receive additional storage capacity or refund current storage capacity.
+    pub resource interface StorageReservationReceiver {
+        pub fun deposit(from: @FlowToken.Vault) {
+            pre {
+                self.isInstance(Type<@StorageReservation>()): "The interface StorageReservationReceiver should only point to resource StorageReservation"
```

Comment:
Nice! Good use of `isInstance`! I haven't found a place to use it yet, but this is perfect